File: hospital-frontend/src/features/auth/services/authService.ts
/**
 * @module features/auth/services/authService
 *
 * Service layer untuk autentikasi pengguna (registrasi & login) ke
 * endpoint `/auth/*` pada backend Spring Boot. Setiap fungsi
 * mengembalikan `AuthResponse` berisi JWT token dan data user yang
 * kemudian disimpan oleh {@link module:core/context/AuthContext}.
 */
import api from "../../../core/api/client";
import type { User } from "../../../core/types";

/**
 * Payload untuk registrasi akun baru.
 */
export interface RegisterPayload {
  /** Nama lengkap pengguna */
  nama: string;
  /** Email pengguna (dipakai sebagai identitas login) */
  email: string;
  /** Password mentah (di-hash di sisi backend) */
  password: string;
}

/**
 * Payload untuk login pengguna.
 */
export interface LoginPayload {
  /** Email terdaftar */
  email: string;
  /** Password mentah */
  password: string;
}

/**
 * Respons autentikasi dari backend setelah login/register berhasil.
 */
export interface AuthResponse {
  /** JWT token yang dipakai untuk request selanjutnya */
  token: string;
  /** Data user yang sedang login */
  user: User;
}

/**
 * Mendaftarkan akun pengguna baru ke backend.
 *
 * @param data - data registrasi (nama, email, password)
 * @returns Promise Axios berisi {@link AuthResponse} (token + user)
 * @throws AxiosError - jika email sudah terpakai atau validasi gagal (mis. 400/409)
 *
 * @example
 * const res = await registerUser({ nama: "Budi", email: "budi@mail.com", password: "rahasia" });
 * login(res.data.token, res.data.user);
 */
export const registerUser = (data: RegisterPayload) =>
  api.post<AuthResponse>("/auth/register", data);

/**
 * Melakukan login pengguna dan memperoleh JWT token.
 *
 * @param data - kredensial login (email, password)
 * @returns Promise Axios berisi {@link AuthResponse} (token + user)
 * @throws AxiosError - jika kredensial salah (mis. 401)
 *
 * @example
 * const res = await loginUser({ email: "budi@mail.com", password: "rahasia" });
 * console.log(res.data.user.nama);
 */
export const loginUser = (data: LoginPayload) =>
  api.post<AuthResponse>("/auth/login", data);

File: medikasentosa/src/main/java/com/medikasentosa/features/auth/service/AuthService.java
package com.medikasentosa.features.auth.service;

import com.medikasentosa.features.auth.dto.*;
import com.medikasentosa.features.auth.entity.Role;
import com.medikasentosa.features.auth.entity.User;
import com.medikasentosa.shared.exception.DuplicateResourceException;
import com.medikasentosa.shared.exception.ResourceNotFoundException;
import com.medikasentosa.features.auth.repository.UserRepository;
import com.medikasentosa.shared.security.JwtService;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

/**
 * Service untuk logika bisnis autentikasi pengguna.
 * Bertanggung jawab atas registrasi user baru, proses login,
 * hashing password, dan pembuatan token JWT.
 *
 * @author Ari
 * @since 1.0.0
 */
@Service
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private JwtService jwtService;

    public AuthService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtService jwtService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtService = jwtService;
    }

    /**
     * Mendaftarkan pengguna baru: memvalidasi keunikan email, meng-hash password,
     * menyimpan user dengan role PASIEN, lalu membuat token JWT.
     *
     * @param request data registrasi (nama, email, password)
     * @return token JWT beserta data user yang baru dibuat
     * @throws DuplicateResourceException bila email sudah terdaftar
     */
    // Register
    public AuthResponse register(RegisterRequest request) {

        // Validasi cek email sudah dipakai atau belum
        if (userRepository.existsByEmail(request.email())) {
            throw new DuplicateResourceException("Email sudah terdaftar");
        }

        // Membuat user
        User user = User.builder().nama(request.nama())
                .email(request.email()).password(passwordEncoder.encode(request.password()))
                .role(Role.PASIEN).isActive(true).build();

        userRepository.save(user);

        // Generate token setelah berhasil save user
        String token = jwtService.generateToken(user.getEmail());

        return new AuthResponse(token, toUserResponse(user));
    }

    /**
     * Memproses login: mencari user berdasarkan email dan memverifikasi password,
     * kemudian membuat token JWT bila kredensial cocok.
     *
     * @param request kredensial login (email, password)
     * @return token JWT beserta data user
     * @throws ResourceNotFoundException bila email tidak ditemukan atau password salah
     */
    public AuthResponse login(LoginRequest request) {

        // Validasi email apakah sudah ada atau belum
        User user = userRepository.findByEmail(request.email())
                .orElseThrow(() -> new ResourceNotFoundException("Email atau password salah!"));

        // Validasi apakah password cocok
        if (!passwordEncoder.matches(request.password(), user.getPassword())) {
            throw new ResourceNotFoundException("Email atau password salah");
        }

        // Validasi akun masih aktif
        if (Boolean.FALSE.equals(user.getIsActive())) {
            throw new ResourceNotFoundException("Akun Anda dinonaktifkan. Hubungi admin.");
        }

        // Generate token
        String token = jwtService.generateToken(user.getEmail());

        return new AuthResponse(token, toUserResponse(user));
    }

    // Helper user entity userresponse
    private UserResponse toUserResponse(User user) {
        return new UserResponse(
                user.getId(),
                user.getNama(),
                user.getEmail(),
                user.getRole());
    }
}
